src/dune_rules/source_deps.ml

```diff
@@ -0,0 +1,21 @@
+open Import
+
+include Dep.Set.Source_tree (struct
+  module Map_reduce =
+    Source_tree.Dir.Make_map_reduce (Memo) (Monoid.Union (Dep.Set))
+
+  type 'a result = 'a Memo.t
+
+  let union_all dir ~f =
+    let prefix_with, dir = Path.extract_build_context_dir_exn dir in
+    let open Memo.O in
+    Source_tree.find_dir dir >>= function
+    | None -> Memo.return Dep.Set.empty
+    | Some dir ->
+      Map_reduce.map_reduce dir ~traverse:Sub_dirs.Status.Set.all ~f:(fun dir ->
+          let path =
+            Path.append_source prefix_with @@ Source_tree.Dir.path dir
+          in
+          let files = Source_tree.Dir.files dir in
+          Memo.return @@ f ~path ~files)
```

Comment:
Looks like this is another example of a traversal that doesn't need to be in `Memo.t`?

Reply:
(And if we drop `Memo` here, we can get rid of `type 'a result` too.)

Reply:
I think it does. Every monad for `Source_tree.Dir.Make_map_reduce` needs to have `val of_memo : 'a memo -> 'a t`.

Reply:
Hmm, OK. Looks a little fishy though.